src/main/java/org/embulk/gradle/embulk_plugins/ProspectiveDependencies.java

```diff
@@ -0,0 +1,368 @@
+            } catch (final UnexpectedDependencyException ex) {
+                this.compileException = ex;
+            }
+            return this;
+        }
+
+        Builder addRuntimeClasspath(final Set<ResolvedArtifact> runtimeClasspath) {
+            if (this.runtimeVersionMap != null || this.runtimeException != null) {
+                throw new IllegalStateException(
+                        "ProspectiveDependencies.Builder.addRuntimeClasspath is called twice unexpectedly.");
+            }
+            try {
+                this.runtimeVersionMap = buildVersionMapFromResolvedArtifacts(runtimeClasspath);
+            } catch (final UnexpectedDependencyException ex) {
+                this.runtimeException = ex;
+            }
+            return this;
+        }
+
+        ProspectiveDependencies build() throws GradleException {
+            if (this.compileException != null || this.runtimeException != null) {
+                throw new GradleException();  // TODO:
+            }
+            if (this.compileVersionMap == null || this.runtimeVersionMap == null) {
+                throw new NullPointerException(
+                        "ProspectiveDependencies is built without compileClasspath or runtimeClasspath unexpectedly.");
+            }
+            return ProspectiveDependencies.buildFromVersionMaps(this.compileVersionMap, this.runtimeVersionMap);
+        }
+
+        private LinkedHashMap<VersionlessDependency, String> compileVersionMap;
+        private LinkedHashMap<VersionlessDependency, String> runtimeVersionMap;
+
+        private UnexpectedDependencyException compileException;
+        private UnexpectedDependencyException runtimeException;
+    }
+
+    private static class UnexpectedDependencyException extends Exception {
+        UnexpectedDependencyException(
+                final LinkedHashSet<VersionlessDependency> duplicates,
+                final ArrayList<LibraryBinaryIdentifier> libs,
+                final ArrayList<ComponentIdentifier> others) {
+            super(buildMessage(duplicates, libs, others));
+            this.duplicates = duplicates;
+            this.libs = libs;
+            this.others = others;
+        }
+
+        private static String buildMessage(
+                final LinkedHashSet<VersionlessDependency> duplicates,
+                final ArrayList<LibraryBinaryIdentifier> libs,
+                final ArrayList<ComponentIdentifier> others) {
+            final StringBuilder exceptionMessage = new StringBuilder();
+
+            if (!duplicates.isEmpty()) {
+                exceptionMessage.append("Dependency duplicates are found: ");
+                exceptionMessage.append(
+                        duplicates.stream().map(VersionlessDependency::toString).collect(Collectors.joining(", ", "[", "]")));
+            }
+
+            if (!libs.isEmpty()) {
+                if (exceptionMessage.length() > 0) {
+                    exceptionMessage.append(" ");
```

Comment:
nitpick, wondering if separating 2 clauses with space is good enough? It looks a bit difficult to distinguish though.

Reply:
Yeah, I agree that it wouldn't be very readable, but :

* It's an exception message, then it should be in a single line (no newline).
* It's possible to structure it by parenthesis even in a single line, but I thought that such a structure could make it less readable.
* `libs` and `others` are unlikely to happen in reality. I left it here just to leave some information (for ourselves reported via tickets/Issues) when it indeed happens by any chance. No need to be very readable, and it's enough if ourselves can identify it.

Reply:
> I left it here just to leave some information (for ourselves reported via tickets/Issues) when it indeed happens by any chance. No need to be very readable, and it's enough if ourselves can identify it.

It sounds reasonable, thanks for share the detail.